aiida/control/code.py

```diff
@@ -60,17 +26,31 @@ def new(self):
 
         from aiida.orm import Code
 
-        if self.code_type == self.CodeType.STORE_AND_UPLOAD:
-            file_list = [os.path.realpath(os.path.join(self.code_folder, f)) for f in os.listdir(self.code_folder)]
-            code = Code(local_executable=self.code_rel_path, files=file_list)
+        # Will be used at the end to check if all keys are known (those that are not None)
+        passed_keys = set([k for k in self._code_spec.keys() if self._code_spec[k] is not None])
+        used = set()
+
+        if self._get_and_count('code_type', used) == self.CodeType.STORE_AND_UPLOAD:
+            file_list = [
+                os.path.realpath(os.path.join(self.code_folder, f))
+                for f in os.listdir(self._get_and_count('code_folder', used))
+            ]
+            code = Code(local_executable=self._get_and_count('code_rel_path', used), files=file_list)
         else:
-            code = Code(remote_computer_exec=(self.computer, self.remote_abs_path))
+            code = Code(
+                remote_computer_exec=(self._get_and_count('computer', used),
+                                      self._get_and_count('remote_abs_path', used)))
+
+        code.label = self._get_and_count('label', used)
+        code.description = self._get_and_count('description', used)
+        code.set_input_plugin_name(self._get_and_count('input_plugin', used).name)
+        code.set_prepend_text(self._get_and_count('prepend_text', used))
+        code.set_append_text(self._get_and_count('append_text', used))
 
-        code.label = self.label
-        code.description = self.description
-        code.set_input_plugin_name(self.input_plugin.name)
-        code.set_prepend_text(self.prepend_text)
-        code.set_append_text(self.append_text)
+        # Complain if there are keys that are passed but not used
+        if passed_keys - used:
```

Comment:
With the addition of this check it becomes necessary to be able to remove the additional keys.
Usecase:

```
ipython >> builder = CodeBuilder()
ipython >> builder.label = ...
# ... lines and lines of generating and adding the other keys
ipython >> builder.extraneous_key = ...
ipython >> builder.new()
CodeValidationError ...
ipython >> # Now the builder object that has been prepared in arduous work has become useless
ipython >> # And we need to start from scratch!
```

Reply:
I will turn this into an issue, seeing as the commit has already been merged.

Reply:
thanks.
I think it's just a matter of implementing `__del__` for attributes and forwarding the `del` call to the internal dictionary?